### PR TITLE
DPIB-496: AI Feature integrations.

### PR DESCRIPTION
**Pull request by mantasmuliar**
- Developer(s): @mantasmuliar 

- Ticket: https://spryker.atlassian.net/browse/DPIB-496

- Release Group: https://release.spryker.com/release-groups/view/5536

- merge: merge

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   OpenAi        | minor(new)                 |                       |

-----------------------------------------

##### Change log

Initial Release

- Introduced OpenAi module to integrate OpenAi API.

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (dmiseev) [inline LICENSE:1]: self-check: check it before release/merging.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# OpenAi Module
+
+This module is a wrapper around   openai-php/client library.
```
Comment (dmiseev) [inline README.md]:
```suggestion
This module is a wrapper around openai-php/client library.
```

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
+{
+    "name": "spryker-eco/open-ai",
+    "type": "library",
+    "description": "OpenAi module",
+    "license": "MIT",
+    "require": {
+        "php": ">=8.1",
+        "openai-php/client": "^0.8.4",
+        "spryker/kernel": "^3.73.0"
```
Comment (dmiseev) [inline composer.json]: transfer dep is missing

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
+<?php
+
+/**
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]: wrong license

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * @var \OpenAI\Client
+     */
+    protected Client $openAiClient;
+
+    /**
+     * @var \SprykerEco\Client\OpenAi\OpenAiConfig
+     */
+    protected OpenAiConfig $config;
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]:
```suggestion
    protected OpenAiConfig $openAiConfig;
```

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    /**
+     * @var string
+     */
+    protected const OPENAI_MESSAGE_ROLE_USER_VALUE = 'user';
+
+    /**
+     * @var \OpenAI\Client
+     */
+    protected Client $openAiClient;
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]: Should be a bridge

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                [
+                    static::OPENAI_MESSAGE_ROLE_KEY => $user,
+                    static::OPENAI_MESSAGE_CONTENT_KEY => $content,
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @param \OpenAI\Responses\Chat\CreateResponse $createResponse
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]: is it possible to use array instead object from external lib?

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+        return $this->buildOpenAiChatResponse($createResponse);
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\OpenAiChatRequestTransfer $openAiRequestTransfer
+     *
+     * @return array<string, mixed>
+     */
+    protected function buildPromptRequest(OpenAiChatRequestTransfer $openAiRequestTransfer): array
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]: Mapper responsiblity

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    protected function buildOpenAiChatResponse(CreateResponse $createResponse): OpenAiChatResponseTransfer
+    {
+        $responseTransfer = new OpenAiChatResponseTransfer();
+        if (!$createResponse->choices) {
+            return $responseTransfer;
+        }
+
+        return $responseTransfer->setMessage(
+            $createResponse->choices[0]->message->content,
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]: also mapper responsibility

File: src/SprykerEco/Client/OpenAi/OpenAiClient.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+namespace SprykerEco\Client\OpenAi;
+
+use Generated\Shared\Transfer\OpenAiChatRequestTransfer;
+use Generated\Shared\Transfer\OpenAiChatResponseTransfer;
+use Spryker\Client\Kernel\AbstractClient;
+
+/**
+ * {@inheritDoc}
+ *
+ * @api
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/OpenAiClient.php]: get rid

File: src/SprykerEco/Client/OpenAi/OpenAiClientInterface.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * @api
+ *
+ * @method \SprykerEco\Client\OpenAi\OpenAiFactory getFactory()
+ */
+interface OpenAiClientInterface
+{
+    /**
+     * Specification:
+     * - Sends a message to ChatGPT engine.
+     * - Returns `OpenAiChatResponseTransfer` object.
```
Comment (dmiseev) [inline src/SprykerEco/Client/OpenAi/OpenAiClientInterface.php:19]: Specify required params.

File: src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php (hunk 6 above)
Comment (mantasmuliar) [reply src/SprykerEco/Client/OpenAi/Chat/OpenAiChat.php]: I'm not sure about it, couldn't find a single bridge example for external dependency. I.e `Spryker\Service\FlysystemAws3v3FileSystem\Model\Builder\Adapter\Aws3v3AdapterBuilder` instantiates S3 Client directly inside the class. I would add an abstraction layer - something like `OpenAiClientProviderPlugin` and then connect it throught DependencyProvider, or straight up instantiate OpenAi `Client` directly inside the dependency provider. But bridge does not seems suitable here 🤔

**Review by dmiseev — APPROVED** [review]

**Review by dmiseev — APPROVED** [review]

**Pull request merged**